### PR TITLE
Improve-template-api

### PR DESCRIPTION
**Pull request by AlfonMnz**
Finish the issue improve template api

### CODE REVIEW

File: src/package.json [hunk 1]
```diff
@@ -4,7 +4,9 @@
   "description": "Template for apis",
   "main": "index.js",
   "scripts": {
-    "start": "node index.js"
+    "start": "node index.js",
+    "dev": "NODE_ENV=run nodemon -L --inspect=0.0.0.0 index.js",
+    "test": "NODE_ENV=test jest --passWithNoTests --maxWorkers=1"
```
Comment (EsteveSegura) [inline src/package.json]: can we add 2 commands?

"pretest" that runs eslint .
and "lint" that runst eslint .

File: src/infrastructure/gracefulStopper.js [hunk 2]
```diff
@@ -0,0 +1,13 @@
+const container = require('../container');
+const signals = ['SIGTERM', 'SIGINT'];
+
+function registerGracefulStopper() {
```
Comment (EsteveSegura) [inline src/infrastructure/gracefulStopper.js]: registerGracefulStopper, its a weird name?
Can we improve that?

File: src/infrastructure/gracefulStopper.js [hunk 3]
```diff
@@ -0,0 +1,13 @@
+const container = require('../container');
+const signals = ['SIGTERM', 'SIGINT'];
```
Comment (EsteveSegura) [inline src/infrastructure/gracefulStopper.js]: add SIGUSR1 and SIGUSR2

File: src/infrastructure/gracefulStopper.js [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+
+function registerGracefulStopper() {
+    const dbHandler = container.resolve('mongoDbHandler');
+    signals.map(signal => process.on(signal, () => {
+        dbHandler.disconnect();
+        process.exit();
```
Comment (EsteveSegura) [inline src/infrastructure/gracefulStopper.js]: process.exit(0);
just to make sure that exists with an OK code.

File: src/infrastructure/gracefulStopper.js [hunk 5]
```diff
@@ -0,0 +1,13 @@
+const container = require('../container');
+const signals = ['SIGTERM', 'SIGINT'];
+
+function registerGracefulStopper() {
+    const dbHandler = container.resolve('mongoDbHandler');
+    signals.map(signal => process.on(signal, () => {
+        dbHandler.disconnect();
```
Comment (EsteveSegura) [inline src/infrastructure/gracefulStopper.js]: Can we do logs after and before of stopping the server? just for having a visual confirmation

File: src/index.js [hunk 6]
```diff
@@ -1,14 +1,29 @@
 require('dotenv').config();
+
+const container = require('./container');
 const express = require('express');
+const gracefulStopper = require('./infrastructure/gracefulStopper');
+gracefulStopper.registerGracefulStopper();
```
Comment (EsteveSegura) [inline src/index.js]: It looks strange to call this method, the execution of the method is obscure.

The feeling is that when this method is called, the server will stop.

We can add the signals and the logic to listen to the process in the index.js and in the gracefulStopper we only leave the logic related to stop the services.

File: README.md [hunk 7]
```diff
@@ -1,2 +1,5 @@
 # template-api
 Template for apis
+
+# Things ToDo
+In db handler check ToDo and set correct collection
```
Comment (EsteveSegura) [inline README.md]: Can we provide a little more information?

Files to edit, lines, fields. Something similar to a checklist, to be sure that we have configured the scaffold correctly.

File: src/index.js [hunk 8]
```diff
@@ -1,14 +1,29 @@
...
 
 const port = process.env.PORT || 3000;
 const secretOrNot = process.env.SECRET || 'NotASecret';
 
+
 app.use('/', (req,res) => {
   res.status(200).json({response: 'OK', secretOrNot});
 })
 
+app.get('/internal/liveness', (req, res) => {
```
Comment (EsteveSegura) [inline src/index.js]: These are controllers,
we'll have to move this into infrastructure/http/internal

File: src/index.js [hunk 9]
```diff
@@ -1,14 +1,29 @@
...
+  res.status(200).json({response: "I'm alive"});
+});
+
+app.get('/internal/readyness', (req, res) => {
+  const dbHandler = container.resolve('mongoDbHandler');
+  if (dbHandler.getInstance()) return res.status(200).json({response: "I'm ready"});
```
Comment (EsteveSegura) [inline src/index.js]: This part of the logic is completely coupled to mongodb, can we make it a bit more agnostic?

File: src/infrastructure/config/index.js [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+const test = {
+  server: {
+    port: 3000
+  },
+  mongo:{
+    mongoConnectionUri: null,
+    mongoTimeout: 1,
+    dbName: 'NOT EXISTS'
+  },
+  keepsecrets:{
```
Comment (EsteveSegura) [inline src/infrastructure/config/index.js]: this is really needed for the scaffold?

**Review by EsteveSegura — CHANGES_REQUESTED** [review]
Algún punto a revisar.

File: src/index.js (hunk 9 above)
Comment (AlfonMnz) [reply src/index.js]: I don't know how want implement it

File: src/infrastructure/config/index.js (hunk 10 above)
Comment (EsteveSegura) [reply src/infrastructure/config/index.js]: Dont remove everythin', just the JWT!

File: src/index.js (hunk 9 above)
Comment (EsteveSegura) [reply src/index.js]: You need to check an array and that everything inside is valid.
Somethign like that

```js
router.get('/readiness', (req, res) => {
    let ready = true;
    const dbHandler = container.resolve('mongoDbHandler');
    const cacheHandler = container.resolve('redisDbHandler');
    const thirdPartyService = container.resolve('thirdParty');

    const services = [dbHandler.getInstance(),cacheHandler.alive(), thirdPartyService.isConnected]
    // loop in services with foreach, and if everyting passes, its ok.
    services.forEach((svc) => {
       if(!svc) {
           ready = false;
       }
    })
    
    if (ready)
          200
     else
          500
});
```

Comment (EsteveSegura) [reply src/index.js]: Also **readyness** is a typo

File: src/infrastructure/http/internal-controller.js [hunk 11]
```diff
@@ -0,0 +1,29 @@
...
+    let ready = true;
+    const dbHandler = container.resolve('mongoDbHandler');
+    try {
+        const services = [await dbHandler.getInstance()];
+        services.forEach((service) => {
+            if (!service) {
+                ready = false
+            }
+        });
+        if (!ready) return res.status(500).json({response: "Not ready"});
```
Comment (EsteveSegura) [inline src/infrastructure/http/internal-controller.js]: avoid oneliners

**Review by EsteveSegura — APPROVED** [review]
Well done! Thanks

**Pull request merged**